# Proyecto Final

### 2019 - II

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co


### Primer integrante:
Nombre: Jorge Hiler Ricardo


#### Segundo integrante:

Nombre: Santiago Gaviria Zapata


In [3]:
import numpy as np
import numpy.matlib as npMatlib
import matplotlib.pyplot as plt
import pandas as pd
from scipy import stats

#Algunas advertencias que queremos evitar
import warnings
warnings.filterwarnings("always")

train = pd.read_csv(r'DB/train.csv')

test = pd.read_csv(r'DB/test.csv')

print(np.size(train, 0), np.size(train, 1))

1460 77


In [4]:


def fillNAs(train, test):
    naTrainColumns = train.columns[train.isna().any()].tolist()
    train[naTrainColumns[0]].fillna(train[naTrainColumns[0]].mean(), inplace = True)
    train[naTrainColumns[1]].fillna('None', inplace = True)
    train[naTrainColumns[2]].fillna(train[naTrainColumns[2]].mean(), inplace = True)
    train[naTrainColumns[3]].fillna('NA', inplace = True)
    train[naTrainColumns[4]].fillna('NA', inplace = True)
    train[naTrainColumns[5]].fillna('NA', inplace = True)
    train[naTrainColumns[6]].fillna('NA', inplace = True)
    train[naTrainColumns[7]].fillna('NA', inplace = True)
    train[naTrainColumns[8]].fillna('SBrkr', inplace = True)
    train[naTrainColumns[9]].fillna('NA', inplace = True)
    train[naTrainColumns[10]].fillna('NA', inplace = True)
    train[naTrainColumns[11]].fillna(train[naTrainColumns[11]].mean(), inplace = True)
    train[naTrainColumns[12]].fillna('NA', inplace = True)
    train[naTrainColumns[13]].fillna('NA', inplace = True)
    train[naTrainColumns[14]].fillna('NA', inplace = True)
    
    naTrainResultantColumns = train.columns[train.isna().any()].size
    print('Numero de columnas con valores NA: ', naTrainResultantColumns)
    
    naTestColumns = test.columns[test.isna().any()].tolist()
    
    test[naTestColumns[0]].fillna('RL', inplace = True)
    test[naTestColumns[1]].fillna(test[naTestColumns[1]].mean(), inplace = True)
    test[naTestColumns[2]].fillna('None', inplace = True)
    test[naTestColumns[3]].fillna(test[naTestColumns[3]].mean(), inplace = True)
    test[naTestColumns[4]].fillna('NA', inplace = True)
    test[naTestColumns[5]].fillna('NA', inplace = True)
    test[naTestColumns[6]].fillna('NA', inplace = True)
    test[naTestColumns[7]].fillna('NA', inplace = True)
    test[naTestColumns[8]].fillna(test[naTestColumns[8]].mean(), inplace = True)
    test[naTestColumns[9]].fillna('NA', inplace = True)
    test[naTestColumns[10]].fillna(test[naTestColumns[10]].mean(), inplace = True)
    test[naTestColumns[11]].fillna(test[naTestColumns[11]].mean(), inplace = True)
    test[naTestColumns[12]].fillna(test[naTestColumns[12]].mean(), inplace = True)
    test[naTestColumns[13]].fillna(test[naTestColumns[13]].mean(), inplace = True)
    test[naTestColumns[14]].fillna(test[naTestColumns[14]].mean(), inplace = True)
    test[naTestColumns[15]].fillna('NA', inplace = True)
    test[naTestColumns[16]].fillna('NA', inplace = True)
    test[naTestColumns[17]].fillna(test[naTestColumns[17]].mean(), inplace = True)
    test[naTestColumns[18]].fillna('NA', inplace = True)
    test[naTestColumns[19]].fillna(round(float(test[naTestColumns[19]].mean()), 1), inplace = True)
    test[naTestColumns[20]].fillna(test[naTestColumns[20]].mean(), inplace = True)
    test[naTestColumns[21]].fillna('NA', inplace = True)
    test[naTestColumns[22]].fillna('NA', inplace = True)
    
    naTestResultantColumns = train.columns[train.isna().any()].size
    print('Numero de columnas con valores NA: ', naTestResultantColumns)
    
    if (naTrainResultantColumns == 0 & naTestResultantColumns == 0):
        return print("Lllenado de campos vacíos completado")
    else: 
        return print("Llenado de campos vacíos erróneo")

fillNAs(train, test)


Numero de columnas con valores NA:  0
Numero de columnas con valores NA:  0
Lllenado de campos vacíos completado


In [6]:

#Deleted columns 'cause had a lot of NA > 80%: 'Alley','Fence','MiscFeature','PoolQC'
#Deleted columns 'cause the std = 0: 'Utilities' %rep = 1, 'Street' %rep =1,
#'LandSlope' %rep = 0.95, Condition2 %rep = 1445/1460, RoofMatl %rep = 1434/1460, 
# Heating %rep = 1428/1460, LowQualFinSF %rep = 1436/1460, 
# PoolArea %rep = 1454/1460, 3SsnPorch %rep = 1438/1460

def deleteColumns(train, test):
    columnsToDelete = ['Id','Utilities','Street','LandSlope', 'Condition2', 'RoofMatl',  'Heating', 'LowQualFinSF', 'PoolArea', '3SsnPorch']
    train.drop(labels = columnsToDelete, axis = 1, inplace = True)
    test.drop(labels = columnsToDelete, axis = 1, inplace = True)
    
deleteColumns(train, test)

KeyError: "['Id' 'Utilities' 'Street' 'LandSlope' 'Condition2' 'RoofMatl' 'Heating'\n 'LowQualFinSF' 'PoolArea' '3SsnPorch'] not found in axis"

In [7]:

def mergeTrainAndTest(train, test):
    trainToMerge = train.copy()
    Ytrain = trainToMerge['SalePrice'] 
    trainToMerge.drop(labels = 'SalePrice', axis = 1, inplace = True)
    trainToMerge['Set'] = 'train'
    testToMerge = test.copy()
    testToMerge['Set'] = 'test'
    db = trainToMerge.copy()
    db = db.append(testToMerge)
    return db, Ytrain

def convertToCategories(db):
    db["MSSubClass"] = db['MSSubClass'].astype('category')
    db["ExterQual"] = db["ExterQual"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1})
    db["ExterCond"] = db["ExterCond"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1})
    db["BsmtQual"] = db["BsmtQual"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1, 'NA': 0})
    db["BsmtCond"] = db["BsmtCond"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1, 'NA': 0})
    db["BsmtExposure"] = db["BsmtExposure"].map({'Gd': 4, 'Av': 3, 'Mn': 2, 'No': 1, 'NA': 0})
    db["HeatingQC"] = db["HeatingQC"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1})
    db["KitchenQual"] = db["KitchenQual"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1})
    db["FireplaceQu"] = db["FireplaceQu"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1, 'NA': 0})
    db["GarageQual"] = db["GarageQual"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1, 'NA': 0})
    db["GarageCond"] = db["GarageCond"].map({'Ex': 5, 'Gd': 4, 'TA': 3, 'Fa': 2, 'Po': 1, 'NA': 0})
    for i in db.select_dtypes(include='object').columns.to_list():
        db[i] = db[i].astype('category')
        
def oneHotEncoding(db):
    typeOfSet = db['Set']
    db.drop(labels = 'Set', axis = 1, inplace = True)
    featuresToEncode = db.select_dtypes(include='category').head().columns.values.tolist()
    resultantDB = pd.get_dummies(db, columns = featuresToEncode, drop_first= True)
    resultantDB = pd.concat([resultantDB, typeOfSet], axis=1)
    return resultantDB

db, Ytrain = mergeTrainAndTest(train, test)
convertToCategories(db)

db = oneHotEncoding(db)

db.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2912 entries, 0 to 1451
Data columns (total 204 columns):
LotFrontage              float64
LotArea                  int64
OverallQual              int64
OverallCond              int64
YearBuilt                int64
YearRemodAdd             int64
MasVnrArea               float64
ExterQual                int64
ExterCond                int64
BsmtQual                 int64
BsmtCond                 int64
BsmtExposure             int64
BsmtFinSF1               float64
BsmtFinSF2               float64
BsmtUnfSF                float64
TotalBsmtSF              float64
HeatingQC                int64
1stFlrSF                 int64
2ndFlrSF                 int64
GrLivArea                int64
BsmtFullBath             float64
BsmtHalfBath             float64
FullBath                 int64
HalfBath                 int64
BedroomAbvGr             int64
KitchenAbvGr             int64
KitchenQual              int64
TotRmsAbvGrd             int64
Fireplac

In [9]:
def separateTestAndTrain(db):
    
    Xtrain = db.loc[db['Set'] == 'train']
    Xtrain.drop(labels = 'Set', axis = 1, inplace = True)
    Xtest = db.loc[db['Set'] == 'test']
    Xtest.drop(labels = 'Set', axis = 1, inplace = True)
    return Xtrain, Xtest

Xtrain, Xtest = separateTestAndTrain(db)

/home/sady/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,
/home/sady/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


# PCA


In [161]:
def get_params(model, params, X_val, y_val):
    estimator = GridSearchCV(estimator = regr_freq, param_grid = regression_grid, cv = 3)
    estimator.fit(X_val, y_val)
    return estimator.best_params_
    

# Regresión lineal


In [90]:
from sklearn.model_selection import train_test_split

#Separa conjunto de entrenamiento y validación del conjunto de test
X_train_val, X_test, y_train_val, y_test = train_test_split(Xtrain, Ytrain, test_size=0.10, shuffle=True)

#Separar entrenamiento del conjunto de validación
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.20, shuffle=True)

In [49]:
#Mejores hiperparametros obtenidos con conjunto de validación
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV
from numpy import random
import math


fit_intercept=[False,True]
normalize=[False,True]
copy_X=[False,True]
n_jobs=[None,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
#scoring = ["neg_mean_absolute_error", "r2"]

regression_grid = {'fit_intercept':fit_intercept, 'normalize':normalize, 'copy_X':copy_X, 'n_jobs':n_jobs}

regr_freq = LinearRegression()
reg_estimator = GridSearchCV(estimator = regr_freq, param_grid = regression_grid, cv = 3)
reg_estimator.fit(X_val, y_val)

reg_estimator.best_params_

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


{'copy_X': False, 'fit_intercept': True, 'n_jobs': None, 'normalize': False}

In [137]:

from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score
from math import sqrt
#Medidas de error

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

def errorMeasure(y_train, y_pred):
    
    rmse = sqrt(mean_squared_error(y_train, y_pred))
    mae = mean_absolute_error(y_train, y_pred)
    mape = mean_absolute_percentage_error(y_train, y_pred)
    r2 = r2_score(y_train, y_pred)
    
    return {'rmse': rmse,'mae': mae, 'mape': mape, 'r2': r2}

In [76]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import StratifiedKFold, GridSearchCV
from numpy import random
import math

regr_freq = LinearRegression(fit_intercept=True)
regr_freq.fit(X_train, y_train)


#Desempeño del modelo con datos de entrenamiento
y_pred_train_LR = regr_freq.predict(X_train)

performanceTrain = errorMeasure(y_train, y_pred_train_LR)
print('performanceTrain', performanceTrain)

#Desempeño del modelo con datos de prueba
y_pred_test_LR = regr_freq.predict(X_test)

performanceTest = errorMeasure(y_test, y_pred_test_LR)
print('performanceTest', performanceTest)



performanceTrain {'rmse': 27518.30242455366, 'mae': 17813.133535132187, 'mape': 10.34361972922708, 'r2': 0.8882679950262543}
performanceTest {'rmse': 26171.140732184904, 'mae': 19506.744503532627, 'mape': 11.622605761568671, 'r2': 0.8368799404253728}


# Ventana de Parzen


In [79]:
#Normalizar los datos

from sklearn import preprocessing
normalize = preprocessing.StandardScaler().fit(X_train)
X_train_N = normalize.transform(X_train)
X_test_N = normalize.transform(X_test)  

normalize = preprocessing.StandardScaler().fit(X_val)
X_val_N = normalize.transform(X_val)

In [84]:
from sklearn.neighbors import KernelDensity
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold, GridSearchCV
from numpy import random
import math

bandwidth = np.logspace(-1, 1, 20)
kernel = ['gaussian']

params = {'bandwidth': bandwidth, 'kernel': kernel}
parzen_grid = GridSearchCV(KernelDensity(), params)
parzen_grid.fit(X_val_N, y_val)

parzen_grid.best_params_

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


{'bandwidth': 0.6951927961775606, 'kernel': 'gaussian'}

In [87]:
from kernel_regression import KernelRegression

regr_parzen = KernelRegression()
regr_parzen.fit(X_train_N, y_train)

#regr_parzen = KernelDensity(algorithm='ball_tree', bandwidth=0.6951927961775606, kernel='gaussian')
#regr_parzen.fit(X_train_N, y_train)

#Desempeño del modelo con datos de entrenamiento
y_pred_train_parzen = regr_parzen.predict(X_train_N)

performanceTrain = errorMeasure(y_train, y_pred_train_parzen)
print('performanceTrain', performanceTrain)

#Desempeño del modelo con datos de prueba
y_pred_test_parzen = regr_parzen.predict(X_test_N)

performanceTest = errorMeasure(y_test, y_pred_test_parzen)
print('performanceTest', performanceTest)

performanceTrain {'rmse': 73186.4056415442, 'mae': 50996.816757912886, 'mape': 31.941266227130953, 'r2': 0.20969423405832088}
performanceTest {'rmse': 56210.4440566955, 'mae': 45490.67436931434, 'mape': 27.851245162307652, 'r2': 0.24751803047668142}


# Regresión Lineal con PCA


In [162]:
i = 0
components = list(range(25,50))
X_pca = np.zeros(75)

for i in components:
    ext = ''
    ext = PCA(n_components=i)
    ext = ext.fit(Xtrain)
    X_ext = ext.transform(Xtrain)
    #Separa conjunto de entrenamiento y validación del conjunto de test
    X_train_val_ex, Xtest_ex, y_train_val_ex, y_test_ex = train_test_split(X_ext, Ytrain, test_size=0.10, shuffle=True)
    
    #Separar entrenamiento del conjunto de validación
    Xtrain_ex, X_val_ex, y_train_ex, y_val_ex = train_test_split(X_train_val_ex, y_train_val_ex, test_size=0.20, shuffle=True)
    
    fit_intercept=[False,True]
    normalize=[False,True]
    copy_X=[False,True]
    n_jobs=[None,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]
    
    regression_grid = {'fit_intercept':fit_intercept, 'normalize':normalize, 'copy_X':copy_X, 'n_jobs':n_jobs}

    
    params = get_params(LinearRegression(), regression_grid, X_val_ex, y_val_ex)

    regr_freq = LinearRegression(fit_intercept = params['fit_intercept'], normalize = params['normalize'],
                                copy_X = params['copy_X'], n_jobs = params['n_jobs'])
    regr_freq.fit(Xtrain_ex, y_train_ex)


    #Desempeño del modelo con datos de entrenamiento
    y_pred_train_LR = regr_freq.predict(Xtrain_ex)

    performanceTrain = errorMeasure(y_train_ex, y_pred_train_LR)
    print('performanceTrain ' + str(i), performanceTrain)

    #Desempeño del modelo con datos de prueba
    y_pred_test_LR = regr_freq.predict(Xtest_ex)

    performanceTest = errorMeasure(y_test_ex, y_pred_test_LR)
    print('performanceTest' + str(i), performanceTest)
    i = i + 1


performanceTrain 25 {'rmse': 2327279.570626905, 'mae': 2325803.259238475, 'mape': 1487.3252746182047, 'r2': -784.7853226658605}
performanceTest25 {'rmse': 25946.382427004097, 'mae': 18914.327220147083, 'mape': 11.154415081851903, 'r2': 0.8694576708122725}
performanceTrain 26 {'rmse': 35882.03091236191, 'mae': 21399.074393061743, 'mape': 12.66494874245821, 'r2': 0.800086001145954}
performanceTest26 {'rmse': 35348.53301606005, 'mae': 24447.75300687702, 'mape': 15.305562930940098, 'r2': 0.8014474610568375}
performanceTrain 27 {'rmse': 36507.55896082327, 'mae': 21668.633204088233, 'mape': 12.769966770259108, 'r2': 0.8017544615395966}
performanceTest27 {'rmse': 28043.184508127124, 'mae': 20581.501733821366, 'mape': 12.159646331029002, 'r2': 0.8656309595241007}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 28 {'rmse': 31183.08641908781, 'mae': 19493.2412046882, 'mape': 11.906862660404322, 'r2': 0.8446851945504239}
performanceTest28 {'rmse': 27177.450459994063, 'mae': 17442.7545219108, 'mape': 11.34100761493539, 'r2': 0.8798914768680598}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 29 {'rmse': 770402.8496836859, 'mae': 766196.3141634096, 'mape': 477.00063901479217, 'r2': -90.47223769432215}
performanceTest29 {'rmse': 22423.93117224636, 'mae': 18226.098969589006, 'mape': 11.96093728459314, 'r2': 0.8826863899666214}
performanceTrain 30 {'rmse': 30200.704067015544, 'mae': 19422.015663568516, 'mape': 11.822160374589618, 'r2': 0.8511567273958673}
performanceTest30 {'rmse': 68215.90634264778, 'mae': 27332.332887883593, 'mape': 15.723904903386288, 'r2': 0.5501787938399327}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 31 {'rmse': 36184.76760415947, 'mae': 21329.55246458575, 'mape': 12.881810453836112, 'r2': 0.796934056482626}
performanceTest31 {'rmse': 30206.615848491994, 'mae': 19479.595018029075, 'mape': 11.400587085515316, 'r2': 0.8204838852670686}
performanceTrain 32 {'rmse': 25705.592666666515, 'mae': 18394.17184863231, 'mape': 11.787157156779832, 'r2': 0.8924837131900711}
performanceTest32 {'rmse': 34019.66020043098, 'mae': 21444.709601509847, 'mape': 11.907262089241916, 'r2': 0.834010827910486}
performanceTrain 33 {'rmse': 34220.56102625905, 'mae': 20095.218658807455, 'mape': 12.03217614077703, 'r2': 0.8035460048216007}
performanceTest33 {'rmse': 22322.115447660763, 'mae': 16977.216722349043, 'mape': 10.705938068893648, 'r2': 0.8980978634312721}
performanceTrain 34 {'rmse': 33621.05932453972, 'mae': 21359.011143682383, 'mape': 12.570687049929374, 'r2': 0.820589226892952}
performanceTest34 {'rmse': 39112.80148029435, 'mae': 22135.516886583242, 'mape': 11.626847496564872, 'r2':

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 36 {'rmse': 1923731.8610961668, 'mae': 1922173.2501849108, 'mape': 1222.3261440510446, 'r2': -612.845212221512}
performanceTest36 {'rmse': 45635.34120407121, 'mae': 23700.005364724886, 'mape': 13.488277198121295, 'r2': 0.7680080212208469}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 37 {'rmse': 26607.65753322739, 'mae': 18504.052138291725, 'mape': 11.372165549536131, 'r2': 0.8865300830134364}
performanceTest37 {'rmse': 22897.85108773663, 'mae': 17388.28578769684, 'mape': 11.381557788786914, 'r2': 0.8313787079113715}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 38 {'rmse': 32965.41302245403, 'mae': 19677.76635702029, 'mape': 11.816969551383947, 'r2': 0.8347674406798493}
performanceTest38 {'rmse': 23740.434586647654, 'mae': 19428.138452306728, 'mape': 12.8245337116339, 'r2': 0.8825956074281481}
performanceTrain 39 {'rmse': 596369.5903857977, 'mae': 595336.7789025758, 'mape': 390.8346678812783, 'r2': -52.63192133385288}
performanceTest39 {'rmse': 26752.472528393406, 'mae': 19661.806225997567, 'mape': 13.55071345253775, 'r2': 0.8579547833294249}
performanceTrain 40 {'rmse': 32705.97324543472, 'mae': 19593.61167262863, 'mape': 11.88707594476169, 'r2': 0.8378165878449235}
performanceTest40 {'rmse': 45300.0768760823, 'mae': 24868.731568257655, 'mape': 13.6049015651939, 'r2': 0.722499416720654}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 41 {'rmse': 1250824.7622318622, 'mae': 1250495.4383026492, 'mape': 813.8947524453399, 'r2': -252.57388098927862}
performanceTest41 {'rmse': 28240.847223930163, 'mae': 18605.290892189718, 'mape': 10.688381546695776, 'r2': 0.8753109382945785}
performanceTrain 42 {'rmse': 1598876.3612003888, 'mae': 1596905.7313457914, 'mape': 1019.0071363111356, 'r2': -402.43552088870547}
performanceTest42 {'rmse': 33910.91001212043, 'mae': 23124.289249612575, 'mape': 13.043742074975064, 'r2': 0.8297039244132915}
performanceTrain 43 {'rmse': 33062.50406493712, 'mae': 19729.37924831241, 'mape': 11.73252167675155, 'r2': 0.8280335626009429}
performanceTest43 {'rmse': 35866.59102345907, 'mae': 22152.335518865988, 'mape': 13.046738315036901, 'r2': 0.8342925082610287}
performanceTrain 44 {'rmse': 32435.389013683976, 'mae': 19798.952268679186, 'mape': 12.169370802323924, 'r2': 0.8308505288293355}
performanceTest44 {'rmse': 39794.49879587556, 'mae': 21014.07388918057, 'mape': 13.02488472462247, '

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 45 {'rmse': 30126.092955132823, 'mae': 18843.398314653947, 'mape': 11.396373574276565, 'r2': 0.8613104009944057}
performanceTest45 {'rmse': 25548.442726127192, 'mae': 18315.39919984398, 'mape': 11.57285144980697, 'r2': 0.8835837689490417}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_search.py:814: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


performanceTrain 46 {'rmse': 27516.328771956174, 'mae': 17761.02217093072, 'mape': 10.673529976035583, 'r2': 0.865236659062515}
performanceTest46 {'rmse': 23733.255735395716, 'mae': 16909.785629478298, 'mape': 10.097581522034906, 'r2': 0.8976447456472971}
performanceTrain 47 {'rmse': 694927.8012101185, 'mae': 690006.44411074, 'mape': 431.6390174057733, 'r2': -69.37979457467277}
performanceTest47 {'rmse': 28642.6092623489, 'mae': 21233.597267538884, 'mape': 12.446612134384416, 'r2': 0.8352623060810955}
performanceTrain 48 {'rmse': 462476.8370949132, 'mae': 455899.27485691383, 'mape': 315.6755973314551, 'r2': -32.0270270372536}
performanceTest48 {'rmse': 25650.149736815376, 'mae': 18366.507645557627, 'mape': 11.043409447827008, 'r2': 0.8939866561051506}
performanceTrain 49 {'rmse': 32293.079229597704, 'mae': 20239.393445361246, 'mape': 11.961495161663866, 'r2': 0.8206983629544057}
performanceTest49 {'rmse': 37052.316108759915, 'mae': 22285.142049862876, 'mape': 12.465583402467564, 'r2': 

In [160]:
from sklearn import preprocessing

i = 0
components = list(range(5,50))
X_pca = np.zeros(75)

for i in components:
    ext = ''
    ext = PCA(n_components=i)
    ext = ext.fit(Xtrain)
    X_ext = ''
    X_ext = ext.transform(Xtrain)
    
    normalize = preprocessing.StandardScaler().fit(X_ext)
    X_ext = normalize.transform(X_ext)
    
    #Separa conjunto de entrenamiento y validación del conjunto de test
    X_train_val_ex, Xtest_ex, y_train_val_ex, y_test_ex = train_test_split(X_ext, Ytrain, test_size=0.10, shuffle=True)
    
    #Separar entrenamiento del conjunto de validación
    Xtrain_ex, X_val_ex, y_train_ex, y_val_ex = train_test_split(X_train_val_ex, y_train_val_ex, test_size=0.20, shuffle=True)
    
    regr_parzen = ''
    regr_parzen = KernelRegression()
    regr_parzen.fit(Xtrain_ex, y_train_ex)

    #Desempeño del modelo con datos de entrenamiento
    y_pred_train_parzen_ex = regr_parzen.predict(Xtrain_ex)

    performanceTrain = errorMeasure(y_train_ex, y_pred_train_parzen_ex)
    print('performanceTrain' + str(i), performanceTrain)

    #Desempeño del modelo con datos de prueba
    y_pred_test_parzen_ex = regr_parzen.predict(Xtest_ex)

    performanceTest = errorMeasure(y_test_ex, y_pred_test_parzen_ex)
    print('performanceTest' + str(i), performanceTest)

performanceTrain5 {'rmse': 63850.98515975949, 'mae': 45020.62884431369, 'mape': 26.594369328165392, 'r2': 0.36502256393478316}
performanceTest5 {'rmse': 63720.550768573354, 'mae': 43138.61704439458, 'mape': 24.371687621032574, 'r2': 0.2762054852651171}
performanceTrain6 {'rmse': 66463.02155833847, 'mae': 46611.686166966065, 'mape': 27.526131558438514, 'r2': 0.31393169076467353}
performanceTest6 {'rmse': 57829.072254239356, 'mae': 43696.987285346426, 'mape': 29.443762176896353, 'r2': 0.2975298933521425}
performanceTrain7 {'rmse': 63042.146917081576, 'mae': 44611.309383723, 'mape': 27.605548760917014, 'r2': 0.2778983272162727}
performanceTest7 {'rmse': 67913.64504737889, 'mae': 50066.43476109931, 'mape': 30.087457414209755, 'r2': 0.20046837350005242}
performanceTrain8 {'rmse': 65874.56738131367, 'mae': 46643.862218228285, 'mape': 28.17219474629809, 'r2': 0.2756032345591346}
performanceTest8 {'rmse': 83443.30621114878, 'mae': 50311.494858839214, 'mape': 30.146901785382376, 'r2': 0.2363559

performanceTrain39 {'rmse': 74711.96270595188, 'mae': 54053.54612024782, 'mape': 33.7790163891315, 'r2': 0.11661873353692687}
performanceTest39 {'rmse': 90579.66717929086, 'mae': 56778.38479590494, 'mape': 35.87448052653667, 'r2': 0.07784417678458067}
performanceTrain40 {'rmse': 78541.01098959039, 'mae': 55385.745343275084, 'mape': 34.315351298962796, 'r2': 0.11066807534293543}
performanceTest40 {'rmse': 65192.82211768962, 'mae': 52223.15942374309, 'mape': 34.210248035539884, 'r2': 0.09662891265406015}
performanceTrain41 {'rmse': 76360.37997498961, 'mae': 54764.26733142069, 'mape': 34.80171966157724, 'r2': 0.11145997187623469}
performanceTest41 {'rmse': 70701.19651249755, 'mae': 50861.33518906018, 'mape': 33.465436823759795, 'r2': 0.09714377977912791}
performanceTrain42 {'rmse': 76841.66314989307, 'mae': 54912.09274353424, 'mape': 34.42200767886503, 'r2': 0.11193274034381095}
performanceTest42 {'rmse': 66035.57354085542, 'mae': 51753.86311711608, 'mape': 35.425625195043956, 'r2': 0.082

# Arboles de decisión con PCA


In [159]:
i = 0
components = list(range(25,70))
X_pca = np.zeros(75)

for i in components:
    ext = ''
    ext = PCA(n_components=i)
    ext = ext.fit(Xtrain)
    X_ext = ''
    X_ext = ext.transform(Xtrain)
    
    normalize = preprocessing.StandardScaler().fit(X_ext)
    X_ext = normalize.transform(X_ext)
    #Separa conjunto de entrenamiento y validación del conjunto de test
    X_train_val_ex, Xtest_ex, y_train_val_ex, y_test_ex = train_test_split(X_ext, Ytrain, test_size=0.10, shuffle=True)
    
    #Separar entrenamiento del conjunto de validación
    Xtrain_ex, X_val_ex, y_train_ex, y_val_ex = train_test_split(X_train_val_ex, y_train_val_ex, test_size=0.20, shuffle=True)
    
    n_estimators = [10, 20, 30, 40, 50, 100,150, 200,300,400,500]
    max_features = ['auto', 'sqrt', 'log2']
    max_depth = [3,4]
    min_samples_split = [0.1]
    criterion = ["mse", "mae"]
    
    random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'criterion': criterion}
    
    params = get_params(RandomForestRegressor(), random_grid, X_val_ex, y_val_ex)
    
    regr = ''
    regr = RandomForestRegressor(n_estimators = params['n_estimators'], max_features = params['max_features'],
                                max_depth = params['max_depth'], min_samples_split = params['min_samples_split'],
                                criterion = params['criterion'])
    regr.fit(Xtrain_ex, y_train_ex)

    Yest = regr.predict(Xtrain_ex) 
    
    performanceTrain = errorMeasure(y_train_ex, Yest)
    print('performanceTrain' + str(i), performanceTrain)

    #Desempeño del modelo con datos de prueba
    Ytest = regr.predict(Xtest_ex)

    performanceTest = errorMeasure(y_test_ex, Ytest)
    print('performanceTest' + str(i), performanceTest)

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain25 {'rmse': 15723.867083299916, 'mae': 9015.34329210276, 'mape': 5.140025224243689, 'r2': 0.9572236501940621}
performanceTest25 {'rmse': 29201.940922986425, 'mae': 19218.95205479452, 'mape': 11.734495624795562, 'r2': 0.8566372099683015}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain26 {'rmse': 15182.351152449965, 'mae': 9428.810180780209, 'mape': 5.344660301584396, 'r2': 0.9623972805373477}
performanceTest26 {'rmse': 35893.55332735911, 'mae': 22632.44109589041, 'mape': 11.884211805244, 'r2': 0.8349137722842375}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain27 {'rmse': 16124.37065272976, 'mae': 9610.347573739295, 'mape': 5.482492398563086, 'r2': 0.9595682666279021}
performanceTest27 {'rmse': 32975.12553322144, 'mae': 21627.744520547945, 'mape': 11.707788546741055, 'r2': 0.7643757641290341}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain28 {'rmse': 18918.531805853145, 'mae': 9472.685061845861, 'mape': 5.310841923498918, 'r2': 0.9406924204832575}
performanceTest28 {'rmse': 44100.71320966627, 'mae': 26494.693150684932, 'mape': 16.410625472757193, 'r2': 0.7754595320006047}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain29 {'rmse': 17404.776604382176, 'mae': 9445.35204567079, 'mape': 5.15896424579308, 'r2': 0.9546559806318691}
performanceTest29 {'rmse': 38887.47084823943, 'mae': 24565.999315068497, 'mape': 14.53007787901249, 'r2': 0.7191938798669196}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain30 {'rmse': 15335.10632525976, 'mae': 9233.717221693625, 'mape': 5.346962283179747, 'r2': 0.9596808886920745}
performanceTest30 {'rmse': 38797.43827431277, 'mae': 26663.071917808218, 'mape': 17.658801132492, 'r2': 0.7575014198222811}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain31 {'rmse': 17261.863764672245, 'mae': 9247.617126546145, 'mape': 5.163137864004116, 'r2': 0.9515102074702162}
performanceTest31 {'rmse': 48415.48062918796, 'mae': 27967.350000000002, 'mape': 15.315087028508339, 'r2': 0.7323286472488545}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain32 {'rmse': 15545.851501723671, 'mae': 9610.802093244529, 'mape': 5.704255805493739, 'r2': 0.9627226898534695}
performanceTest32 {'rmse': 31552.290541214148, 'mae': 21393.37602739726, 'mape': 14.475289197602287, 'r2': 0.8067444406450137}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain33 {'rmse': 15850.211814857303, 'mae': 9495.13368220742, 'mape': 5.3907878126039845, 'r2': 0.9640124719731169}
performanceTest33 {'rmse': 30569.662476145848, 'mae': 21894.22876712329, 'mape': 12.323110963435996, 'r2': 0.7906921818861116}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain34 {'rmse': 17035.20335539685, 'mae': 9317.10875356803, 'mape': 5.311654603440856, 'r2': 0.9557627598412094}
performanceTest34 {'rmse': 40194.130835379125, 'mae': 23827.38698630137, 'mape': 13.49292532563822, 'r2': 0.7314447210646828}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain35 {'rmse': 16976.140483676005, 'mae': 9326.973358705995, 'mape': 5.196034303147459, 'r2': 0.954926180683239}
performanceTest35 {'rmse': 31148.37310993221, 'mae': 22893.28835616438, 'mape': 15.041468290604184, 'r2': 0.8357276808842438}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain36 {'rmse': 16974.692694846966, 'mae': 9483.916079923883, 'mape': 5.422210394228514, 'r2': 0.9545792641142639}
performanceTest36 {'rmse': 42729.938418238315, 'mae': 24208.305479452058, 'mape': 13.026144082858092, 'r2': 0.7215738173465935}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain37 {'rmse': 15799.406881544077, 'mae': 9521.734157944815, 'mape': 5.488354524722698, 'r2': 0.9645779976057175}
performanceTest37 {'rmse': 31442.619976964736, 'mae': 22168.81712328767, 'mape': 13.041211745332612, 'r2': 0.8202419927635531}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain38 {'rmse': 15898.024913870753, 'mae': 9519.017126546147, 'mape': 5.458729289139679, 'r2': 0.958693180036109}
performanceTest38 {'rmse': 53297.12447274256, 'mae': 27924.40273972603, 'mape': 13.524633350678261, 'r2': 0.7210988923219703}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain39 {'rmse': 15949.334318207111, 'mae': 9789.47031398668, 'mape': 5.761977181256988, 'r2': 0.9611434209764917}
performanceTest39 {'rmse': 45254.14981933958, 'mae': 26300.898630136988, 'mape': 15.977017432854812, 'r2': 0.7430391438677559}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain40 {'rmse': 16377.845331839691, 'mae': 9828.372787821123, 'mape': 5.704723144567671, 'r2': 0.9609489105307423}
performanceTest40 {'rmse': 39087.00749056301, 'mae': 24685.452739726028, 'mape': 14.248247414575399, 'r2': 0.6603021605803787}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain41 {'rmse': 18678.675629173402, 'mae': 10102.466317792578, 'mape': 5.676164793808964, 'r2': 0.9463132825693252}
performanceTest41 {'rmse': 25162.031762363815, 'mae': 17648.715753424658, 'mape': 10.880429685206824, 'r2': 0.8633013198669475}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain42 {'rmse': 15463.85672459805, 'mae': 9459.409705042817, 'mape': 5.553143462820287, 'r2': 0.9621348281149769}
performanceTest42 {'rmse': 37414.755072928914, 'mae': 22814.639726027395, 'mape': 13.251505864410587, 'r2': 0.7990715637469786}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain43 {'rmse': 14552.359727407136, 'mae': 8985.121883920077, 'mape': 5.2243317683979935, 'r2': 0.9653341914596743}
performanceTest43 {'rmse': 37196.36121930643, 'mae': 23648.220547945206, 'mape': 12.990967713344933, 'r2': 0.7691185706218365}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain44 {'rmse': 18090.34015687164, 'mae': 9748.5297811608, 'mape': 5.637804713803366, 'r2': 0.9474606795803431}
performanceTest44 {'rmse': 34892.30047135551, 'mae': 22259.484246575343, 'mape': 12.84608313199915, 'r2': 0.7844501648254676}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain45 {'rmse': 16153.095917566183, 'mae': 9565.446431969553, 'mape': 5.477682310804203, 'r2': 0.9599355142463402}
performanceTest45 {'rmse': 49821.17595609472, 'mae': 23225.248630136986, 'mape': 14.066664184584734, 'r2': 0.3237529795779843}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain46 {'rmse': 16206.491860745997, 'mae': 9048.939866793531, 'mape': 5.105233548659349, 'r2': 0.9594746270628824}
performanceTest46 {'rmse': 55827.11737891951, 'mae': 29411.26780821918, 'mape': 14.852424448253524, 'r2': 0.5351838617589697}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain47 {'rmse': 18818.5713208047, 'mae': 9487.314938154139, 'mape': 5.228548842078317, 'r2': 0.9441295724383866}
performanceTest47 {'rmse': 37087.84265764575, 'mae': 25325.041780821917, 'mape': 16.419887073304082, 'r2': 0.7758477060910798}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain48 {'rmse': 17112.534957316402, 'mae': 9653.572121788773, 'mape': 5.512746511245556, 'r2': 0.9547454374022544}
performanceTest48 {'rmse': 33963.9370935223, 'mae': 21723.12328767123, 'mape': 13.955045516688196, 'r2': 0.7269362125104528}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain49 {'rmse': 17057.8668233652, 'mae': 10022.030352045671, 'mape': 5.760131197184308, 'r2': 0.954965659793863}
performanceTest49 {'rmse': 27765.955837539463, 'mae': 20299.563013698633, 'mape': 12.31548022428011, 'r2': 0.8769227919347033}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain50 {'rmse': 15595.961720312898, 'mae': 9096.81560418649, 'mape': 5.321364923178615, 'r2': 0.9586713402044611}
performanceTest50 {'rmse': 43418.20999088913, 'mae': 28409.717123287668, 'mape': 15.250166665465642, 'r2': 0.7624306160010521}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain51 {'rmse': 16554.171259147453, 'mae': 9611.306945765937, 'mape': 5.58113863894611, 'r2': 0.953749313598846}
performanceTest51 {'rmse': 40798.78001593803, 'mae': 26598.471232876713, 'mape': 14.60865809465289, 'r2': 0.7672648524907013}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain52 {'rmse': 17298.297024653606, 'mae': 9867.460418648905, 'mape': 5.553606477892835, 'r2': 0.9520415058974862}
performanceTest52 {'rmse': 43700.68841321344, 'mae': 20586.601369863016, 'mape': 11.941140073971662, 'r2': 0.6756626526056493}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain53 {'rmse': 15877.913758023775, 'mae': 9212.122454804947, 'mape': 5.2686789224368376, 'r2': 0.9555521057136092}
performanceTest53 {'rmse': 51088.291422979164, 'mae': 25718.413013698628, 'mape': 13.652754494706759, 'r2': 0.7419510094782553}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain54 {'rmse': 17481.910975913965, 'mae': 10084.34405328259, 'mape': 5.75087112677577, 'r2': 0.9492642307328956}
performanceTest54 {'rmse': 42787.799048207584, 'mae': 24696.41575342466, 'mape': 13.415449714142833, 'r2': 0.737380107351044}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain55 {'rmse': 18167.79844167258, 'mae': 9715.440723120837, 'mape': 5.604440720506056, 'r2': 0.946763840423259}
performanceTest55 {'rmse': 41040.195746683465, 'mae': 24535.036301369862, 'mape': 14.847023198193288, 'r2': 0.7323673234289669}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain56 {'rmse': 15574.644398644068, 'mae': 9631.532921027592, 'mape': 5.613684659598736, 'r2': 0.9644460251813598}
performanceTest56 {'rmse': 41926.901870666115, 'mae': 22587.498630136986, 'mape': 14.457159313773134, 'r2': 0.5665937274616775}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain57 {'rmse': 18775.057799553248, 'mae': 10046.339771646051, 'mape': 5.61454522808823, 'r2': 0.9456270784471722}
performanceTest57 {'rmse': 38704.6692489739, 'mae': 23615.30821917808, 'mape': 13.338582494313563, 'r2': 0.7478089161374126}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain58 {'rmse': 16842.488055891976, 'mae': 9779.971836346338, 'mape': 5.621377632526898, 'r2': 0.9549316687217266}
performanceTest58 {'rmse': 37061.701886686235, 'mae': 23185.97602739726, 'mape': 13.627039649942684, 'r2': 0.719429079269724}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain59 {'rmse': 16732.5480174665, 'mae': 9420.922930542341, 'mape': 5.556380280805582, 'r2': 0.9512167113601081}
performanceTest59 {'rmse': 45529.91286853328, 'mae': 28147.90205479452, 'mape': 14.19616043681815, 'r2': 0.7689478619276013}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain60 {'rmse': 17375.151700967035, 'mae': 9637.272407231208, 'mape': 5.68886739341178, 'r2': 0.9528022569490074}
performanceTest60 {'rmse': 42988.560663439035, 'mae': 24034.06917808219, 'mape': 14.265421941941907, 'r2': 0.7180270120808617}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain61 {'rmse': 16424.109472216365, 'mae': 9866.648810656516, 'mape': 5.45446641408633, 'r2': 0.9610531085494147}
performanceTest61 {'rmse': 29854.688976518537, 'mae': 20846.84383561644, 'mape': 11.382632784900373, 'r2': 0.8408075235609527}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain62 {'rmse': 15440.601769130728, 'mae': 9435.421313035205, 'mape': 5.389891547688575, 'r2': 0.9599316422387436}
performanceTest62 {'rmse': 49730.588718694504, 'mae': 26825.171232876713, 'mape': 15.118511789160124, 'r2': 0.637159543691225}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain63 {'rmse': 17281.860681249793, 'mae': 9781.474024738345, 'mape': 5.576793855178098, 'r2': 0.9527384091472214}
performanceTest63 {'rmse': 40901.91935122494, 'mae': 25428.038356164383, 'mape': 13.398786828492568, 'r2': 0.7630142385068976}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain64 {'rmse': 18328.733038020386, 'mae': 10213.31274976213, 'mape': 5.775838051351928, 'r2': 0.9486005549734495}
performanceTest64 {'rmse': 37971.01213968794, 'mae': 21530.083561643838, 'mape': 13.896863200332515, 'r2': 0.6686504445099987}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain65 {'rmse': 16810.211611227656, 'mae': 10180.382302568984, 'mape': 5.92731791633609, 'r2': 0.9506937294998872}
performanceTest65 {'rmse': 36220.3523474651, 'mae': 22641.023287671233, 'mape': 14.649162508653891, 'r2': 0.7303606195387682}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain66 {'rmse': 17861.420641456716, 'mae': 9584.68924833492, 'mape': 5.3674187886918565, 'r2': 0.9465456860474947}
performanceTest66 {'rmse': 37262.85168201832, 'mae': 23832.292465753428, 'mape': 14.037211805469235, 'r2': 0.8322868546500667}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain67 {'rmse': 16871.65342663987, 'mae': 9884.484966698381, 'mape': 5.512575503061444, 'r2': 0.9533846428577862}
performanceTest67 {'rmse': 34223.167553859195, 'mae': 23614.56917808219, 'mape': 12.870881302317006, 'r2': 0.8359448125420045}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain68 {'rmse': 16945.25262199407, 'mae': 9881.601522359657, 'mape': 5.7287723710543945, 'r2': 0.9500908185689642}
performanceTest68 {'rmse': 35642.5073283167, 'mae': 23076.356849315067, 'mape': 13.59064935111787, 'r2': 0.8252061493225458}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


performanceTrain69 {'rmse': 18455.92283432852, 'mae': 9390.297145575643, 'mape': 5.504350892237144, 'r2': 0.9479286245728202}
performanceTest69 {'rmse': 34409.640769733756, 'mae': 23477.71712328767, 'mape': 13.790518916850214, 'r2': 0.7888987864781539}


# Regresion por Vectores de Soporte con kernel lineal y con kernel RBF.

In [158]:
from sklearn.svm import SVR

i = 0
components = list(range(25,70))
X_pca = np.zeros(75)

for i in components:
    ext = ''
    ext = PCA(n_components=i)
    ext = ext.fit(Xtrain)
    X_ext = ''
    X_ext = ext.transform(Xtrain)
    
    normalize = preprocessing.StandardScaler().fit(X_ext)
    X_ext = normalize.transform(X_ext)
    
    
    #Separa conjunto de entrenamiento y validación del conjunto de test
    X_train_val_ex, Xtest_ex, y_train_val_ex, y_test_ex = train_test_split(X_ext, Ytrain, test_size=0.10, shuffle=True)
    
    #Separar entrenamiento del conjunto de validación
    Xtrain_ex, X_val_ex, y_train_ex, y_val_ex = train_test_split(X_train_val_ex, y_train_val_ex, test_size=0.20, shuffle=True)
    
    #normalize = preprocessing.StandardScaler().fit(Xtrain_ex)
    #X_train_ex_N = normalize.transform(Xtrain_ex)
    #X_test_ex_N = normalize.transform(Xtest_ex)  
    regr = ''
    regr = SVR()
    regr.fit(Xtrain_ex, y_train_ex)

    Yest = regr.predict(Xtrain_ex)
    
    performanceTrain = errorMeasure(y_train_ex, Yest)
    print('performanceTrain' + str(i), performanceTrain)

    #Desempeño del modelo con datos de prueba
    Ytest = regr.predict(Xtest_ex)

    performanceTest = errorMeasure(y_test_ex, Ytest)
    print('performanceTest' + str(i), performanceTest)

/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain25 {'rmse': 84339.94782289107, 'mae': 56802.38324471701, 'mape': 32.298707865620194, 'r2': -0.04860577880022254}
performanceTest25 {'rmse': 71689.48390729539, 'mae': 54240.680908734335, 'mape': 33.85264731346098, 'r2': -0.03900964463803569}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain26 {'rmse': 77026.36515580403, 'mae': 54621.405918755125, 'mape': 32.4977638334836, 'r2': -0.04036410201766838}
performanceTest26 {'rmse': 79198.41574251371, 'mae': 53269.77470008302, 'mape': 31.851838070917765, 'r2': -0.0143295263153711}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain27 {'rmse': 81264.49074652609, 'mae': 53844.58839377934, 'mape': 30.33959249751959, 'r2': -0.05267439138809826}
performanceTest27 {'rmse': 72289.12791100999, 'mae': 53514.16301760741, 'mape': 30.758632577589335, 'r2': -0.07732473255537808}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain28 {'rmse': 84316.05291420419, 'mae': 56654.707008910664, 'mape': 32.44115668782135, 'r2': -0.048539479234763094}
performanceTest28 {'rmse': 77986.59561703166, 'mae': 52459.93231316972, 'mape': 26.909511734893965, 'r2': -0.07080685743108184}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain29 {'rmse': 78717.32487415794, 'mae': 54291.16175078315, 'mape': 31.138486320152616, 'r2': -0.05150511212533182}
performanceTest29 {'rmse': 83562.92372774798, 'mae': 55038.88639300415, 'mape': 30.979456319557634, 'r2': -0.07065633487643708}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain30 {'rmse': 82701.85246177259, 'mae': 55623.31907248437, 'mape': 31.601309008112217, 'r2': -0.054753605390768234}
performanceTest30 {'rmse': 78516.1992658606, 'mae': 57815.035697627995, 'mape': 35.1999054004825, 'r2': -0.052155865770348475}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain31 {'rmse': 83451.10484568533, 'mae': 57133.71229129671, 'mape': 32.82011214684603, 'r2': -0.05318326770443815}
performanceTest31 {'rmse': 75405.19657103399, 'mae': 51964.541151194484, 'mape': 29.20358484796034, 'r2': -0.0404837372574518}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain32 {'rmse': 81451.7271748908, 'mae': 55665.14534168948, 'mape': 31.47178626353516, 'r2': -0.05208788710214396}
performanceTest32 {'rmse': 82451.23906912115, 'mae': 58312.86540096419, 'mape': 33.606608727383126, 'r2': -0.04114880468084059}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain33 {'rmse': 84501.90558476975, 'mae': 56584.284941342434, 'mape': 31.95938993532871, 'r2': -0.0495347540928015}
performanceTest33 {'rmse': 73326.33994041957, 'mae': 54088.29470597665, 'mape': 29.211044390568823, 'r2': -0.09628455450191775}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain34 {'rmse': 79037.68414199534, 'mae': 54159.67330512352, 'mape': 31.246776302079667, 'r2': -0.04433198443417474}
performanceTest34 {'rmse': 92030.70335502808, 'mae': 61268.36691351995, 'mape': 31.264410948254845, 'r2': -0.0918442544666449}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain35 {'rmse': 80023.06536966357, 'mae': 54568.61840233492, 'mape': 31.406852094594857, 'r2': -0.052235577766430286}
performanceTest35 {'rmse': 74432.1953443766, 'mae': 51202.38810148006, 'mape': 27.606580203404405, 'r2': -0.05965916417268757}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain36 {'rmse': 81462.72730243298, 'mae': 54959.55447242071, 'mape': 31.479390178157047, 'r2': -0.050494050282182235}
performanceTest36 {'rmse': 85738.88761558475, 'mae': 63786.856852076795, 'mape': 34.33176086184946, 'r2': -0.14359096298157548}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain37 {'rmse': 82739.19664442833, 'mae': 55998.56204176766, 'mape': 32.28423221358987, 'r2': -0.05143133315430792}
performanceTest37 {'rmse': 74837.14136898547, 'mae': 53219.14456342094, 'mape': 28.00723725947836, 'r2': -0.08640705638218038}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain38 {'rmse': 82782.22475970314, 'mae': 56386.086326432654, 'mape': 31.359532635585236, 'r2': -0.05827712551001962}
performanceTest38 {'rmse': 64558.21348007447, 'mae': 48105.819546442784, 'mape': 31.489591873773588, 'r2': -0.016617908676919146}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain39 {'rmse': 77982.25741263962, 'mae': 53719.68264841535, 'mape': 31.2274182478286, 'r2': -0.04898499300964976}
performanceTest39 {'rmse': 89363.5163642334, 'mae': 58505.694583930155, 'mape': 32.03973431808201, 'r2': -0.07479513538689608}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain40 {'rmse': 79942.83083526259, 'mae': 55126.15327027038, 'mape': 31.122741704873974, 'r2': -0.05050561893780681}
performanceTest40 {'rmse': 71359.5521312676, 'mae': 52183.48996129218, 'mape': 29.56564795902355, 'r2': -0.05955058776461852}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain41 {'rmse': 83121.81669672433, 'mae': 55836.10155984939, 'mape': 31.511713883773535, 'r2': -0.05458787831330625}
performanceTest41 {'rmse': 76262.3525383858, 'mae': 54151.485213926295, 'mape': 31.70953469624304, 'r2': -0.04269314136074098}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain42 {'rmse': 75066.29810439075, 'mae': 52730.958910051944, 'mape': 31.689381076475488, 'r2': -0.03473897789970648}
performanceTest42 {'rmse': 96747.74242210656, 'mae': 65128.99245348832, 'mape': 31.984233643151818, 'r2': -0.11132737170135298}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain43 {'rmse': 81182.31019083493, 'mae': 55289.87790200605, 'mape': 31.36410460632587, 'r2': -0.058995754295104064}
performanceTest43 {'rmse': 100760.55721260271, 'mae': 62361.0709853336, 'mape': 31.43683771156997, 'r2': -0.1049685258109554}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain44 {'rmse': 74562.09068910066, 'mae': 53033.860478473696, 'mape': 31.61971791181808, 'r2': -0.04299463891711053}
performanceTest44 {'rmse': 111874.15263038862, 'mae': 69607.21093709479, 'mape': 29.69363249026946, 'r2': -0.15993595793858928}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain45 {'rmse': 79070.0273374807, 'mae': 54165.90404413696, 'mape': 30.494457277215602, 'r2': -0.05016875500939544}
performanceTest45 {'rmse': 94765.58179703866, 'mae': 61551.28996794158, 'mape': 33.43171767962282, 'r2': -0.08712155980655956}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain46 {'rmse': 80340.52369977401, 'mae': 54994.85180228637, 'mape': 31.63685827541858, 'r2': -0.04987987751693668}
performanceTest46 {'rmse': 83098.00941503647, 'mae': 57145.12034969674, 'mape': 32.16459029720911, 'r2': -0.04281214616914508}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain47 {'rmse': 81362.81469902182, 'mae': 56331.814364749785, 'mape': 32.72656942855562, 'r2': -0.037855649991850315}
performanceTest47 {'rmse': 73522.10969076087, 'mae': 51052.1208701442, 'mape': 31.160173860645973, 'r2': -0.0029654785122539273}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain48 {'rmse': 80649.68762134793, 'mae': 54921.0535009596, 'mape': 31.00144484318172, 'r2': -0.05037390968319255}
performanceTest48 {'rmse': 86789.38764275551, 'mae': 58904.60638756105, 'mape': 37.721603874509604, 'r2': -0.02067383555588065}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain49 {'rmse': 81971.34335602268, 'mae': 55341.34425672547, 'mape': 31.560823335220928, 'r2': -0.057853162139506864}
performanceTest49 {'rmse': 83461.23055639128, 'mae': 51530.445889512725, 'mape': 25.77322404404225, 'r2': -0.0896751416535706}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain50 {'rmse': 81088.05314457709, 'mae': 55489.09488254349, 'mape': 31.23787912146183, 'r2': -0.061397746352744864}
performanceTest50 {'rmse': 65961.27216351587, 'mae': 47543.389644638075, 'mape': 26.449840191099337, 'r2': -0.10615602494298715}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain51 {'rmse': 80647.36328158059, 'mae': 55407.705895829866, 'mape': 31.50057517613783, 'r2': -0.05113772195114663}
performanceTest51 {'rmse': 101605.0409290193, 'mae': 64176.07529575434, 'mape': 30.442617170435586, 'r2': -0.11535812431953762}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain52 {'rmse': 81871.82881040055, 'mae': 55950.243295316875, 'mape': 31.856672168122756, 'r2': -0.045856359354149445}
performanceTest52 {'rmse': 99235.59978676472, 'mae': 63731.41517415382, 'mape': 32.48783407872412, 'r2': -0.059012980590889885}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain53 {'rmse': 82913.73423636334, 'mae': 55702.77206347557, 'mape': 31.000013908252832, 'r2': -0.06541901012229934}
performanceTest53 {'rmse': 90806.32041243956, 'mae': 60577.8887323174, 'mape': 32.57433880382895, 'r2': -0.0703618015501486}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain54 {'rmse': 80140.98344817357, 'mae': 54443.986601188735, 'mape': 30.37564391662237, 'r2': -0.06430099455099803}
performanceTest54 {'rmse': 72083.16378111049, 'mae': 53609.74016844696, 'mape': 34.10422459222555, 'r2': -0.0636034557916243}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain55 {'rmse': 84372.81674470587, 'mae': 56449.72544366408, 'mape': 31.808715054261278, 'r2': -0.05445771316590031}
performanceTest55 {'rmse': 74051.50625401663, 'mae': 54120.66775844341, 'mape': 33.07070829672132, 'r2': -0.025288707721646375}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain56 {'rmse': 81619.02676767542, 'mae': 54858.97750603263, 'mape': 31.133039735012574, 'r2': -0.04729427493296412}
performanceTest56 {'rmse': 85568.14543073552, 'mae': 60115.19206512249, 'mape': 35.19018687845792, 'r2': -0.052078693152177946}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain57 {'rmse': 77883.34620171733, 'mae': 54173.20443336019, 'mape': 32.05839651691037, 'r2': -0.03946043918749975}
performanceTest57 {'rmse': 81306.55769335291, 'mae': 57234.45924332531, 'mape': 33.115104733144086, 'r2': -0.04521651996516973}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain58 {'rmse': 80640.23890007182, 'mae': 57055.48473202619, 'mape': 32.39992806444869, 'r2': -0.04861116917634911}
performanceTest58 {'rmse': 65140.74986805674, 'mae': 44969.00731856355, 'mape': 27.983133503744128, 'r2': 0.00012722214141513977}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain59 {'rmse': 78122.2087749256, 'mae': 53389.84166498259, 'mape': 31.03227749975874, 'r2': -0.03828839108188076}
performanceTest59 {'rmse': 98124.5552585434, 'mae': 65943.30917545063, 'mape': 36.25774037634532, 'r2': -0.08703057926694036}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain60 {'rmse': 82833.26292741092, 'mae': 55583.22143441504, 'mape': 31.14472354373161, 'r2': -0.05412831572525967}
performanceTest60 {'rmse': 77797.96888051009, 'mae': 57525.33717921745, 'mape': 34.104405749361824, 'r2': -0.04984657099981615}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain61 {'rmse': 79929.51930605365, 'mae': 55242.41433964369, 'mape': 31.52524474588903, 'r2': -0.04716381244673862}
performanceTest61 {'rmse': 88662.38266095487, 'mae': 62480.02237756068, 'mape': 36.11811964982806, 'r2': -0.059161549867966734}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain62 {'rmse': 82257.03527228409, 'mae': 55904.281838087656, 'mape': 32.17786535437531, 'r2': -0.047135471891672776}
performanceTest62 {'rmse': 74589.43338077466, 'mae': 55344.57281488888, 'mape': 32.102836397241944, 'r2': -0.05443451515332631}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain63 {'rmse': 83284.23577855574, 'mae': 56197.48965271221, 'mape': 32.5078661037645, 'r2': -0.04823365455776507}
performanceTest63 {'rmse': 76710.66152408485, 'mae': 56097.425153237324, 'mape': 30.207836131032934, 'r2': -0.09001468550719549}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain64 {'rmse': 80176.62074634118, 'mae': 55259.05382337962, 'mape': 31.027126579276736, 'r2': -0.049588612076582717}
performanceTest64 {'rmse': 72418.27371825029, 'mae': 50894.00901475938, 'mape': 30.26623405707066, 'r2': -0.024036599448968854}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain65 {'rmse': 83230.24452978991, 'mae': 56057.18997436335, 'mape': 31.642748191840163, 'r2': -0.05234449995280621}
performanceTest65 {'rmse': 69439.98732438407, 'mae': 53589.3809754127, 'mape': 35.58919380467743, 'r2': -0.0058704347879647845}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain66 {'rmse': 82330.26557000949, 'mae': 56043.72962020124, 'mape': 32.55109440545469, 'r2': -0.0483609502992246}
performanceTest66 {'rmse': 75092.59247018876, 'mae': 53270.13010966033, 'mape': 27.04913091324883, 'r2': -0.13350276642015668}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain67 {'rmse': 83794.8507665812, 'mae': 56312.080324518975, 'mape': 30.970787665862726, 'r2': -0.061153953318093146}
performanceTest67 {'rmse': 66308.71976522314, 'mae': 48482.3041996222, 'mape': 29.526149678982833, 'r2': -0.07400905115477197}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain68 {'rmse': 78726.3669368458, 'mae': 55154.963904678116, 'mape': 32.56108046008174, 'r2': -0.04085427652575757}
performanceTest68 {'rmse': 67778.4552448964, 'mae': 50905.52218703198, 'mape': 29.202609950320845, 'r2': -0.0431321397887805}


/home/sady/anaconda3/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


performanceTrain69 {'rmse': 83103.61662147322, 'mae': 55483.85945433483, 'mape': 31.594419940139325, 'r2': -0.04469933323660502}
performanceTest69 {'rmse': 70797.7811526137, 'mae': 53022.61845019122, 'mape': 27.820727907984093, 'r2': -0.08963560962070893}


Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [ ]:
qgrid_widget.get_changed_df()

## Ejercicio 5: Pruebas

En la celda de código del Ejercicio # 3, comente la línea donde se normalizan las matrices de datos Xtrain y Xtest. Realice pruebas para diferentes valores de $\eta$ y de grado del polinomio de manera similar a los valores que usó en el punto 3. Observe que pasa con el ECM.

#### Responda

5.1 ¿Qué sucede con los valores del ECM?



Los valores del ECM se desbordan

5.2 ¿A qué se debe lo que observa?


Debido a que la convergencia entre caracteristicas se reduce por falta de normalización

5.3.1 ¿Qu&eacute; proceso hace la normalizaci&oacute;n sobre los datos? 




Cambia los valores del dataset a una escala común sin distorsionar las diferencias en los rangos de valores.

5.3.2 Consulte por qu&eacute; es necesaria la normalizaci&oacute;n en el modelo de regresi&oacute;n  y cu&aacute;les son los tipos de normalizaci&oacute;n m&aacute;s comunes. 

Es necesaria porque es posible que algunas características sean cuantitativamente mucho mayores que las demás lo que puede causar que contribuyan a la variación en el término independiente aunque no necesariamente sean más importantes. Algunos de los métodos de normalización son: decimal scaling, min-max normalization y z-score.

5.3.3 ¿Cu&aacute;l de ellos se aplic&oacute; en el laboratorio?

El método utilizado en este laboratorio fue z-score. 
